modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/consensus/ConsensusStorage.scala

```diff
@@ -0,0 +1,222 @@
+
+  def addFacility(peerId: PeerId, key: Key, bound: Bound): F[ConsensusResources[Artifact]]
+
+  def addProposal(peerId: PeerId, key: Key, hash: Hash): F[ConsensusResources[Artifact]]
+
+  def addSignature(peerId: PeerId, key: Key, signature: Signature): F[ConsensusResources[Artifact]]
+
+  def setLastKeyAndArtifact(value: Option[(Key, Artifact)]): F[Unit]
+
+  def getLastKeyAndArtifact: F[Option[(Key, Artifact)]]
+
+  def tryUpdateLastKeyAndArtifactWithCleanup(oldValue: (Key, Artifact), newValue: (Key, Artifact)): F[Boolean]
+
+}
+
+object ConsensusStorage {
+
+  def make[F[_]: Async: KryoSerializer, Event, Key: Show: Next: Eq, Artifact <: AnyRef: Show: Eq](
+    lastKeyAndArtifact: Option[(Key, Artifact)] = none[(Key, Artifact)]
+  ): F[ConsensusStorage[F, Event, Key, Artifact]] =
+    for {
+      stateUpdateSemaphore <- Semaphore[F](1)
+      lastKeyAndArtifactR <- Ref.of(lastKeyAndArtifact)
+      eventsR <- MapRef.ofConcurrentHashMap[F, PeerId, List[(Ordinal, Event)]]()
+      statesR <- MapRef.ofConcurrentHashMap[F, Key, ConsensusState[Key, Artifact]]()
+      resourcesR <- MapRef.ofConcurrentHashMap[F, Key, ConsensusResources[Artifact]]()
+    } yield make(stateUpdateSemaphore, lastKeyAndArtifactR, eventsR, statesR, resourcesR)
+
+  def make[F[_]: Async: KryoSerializer, Event, Key: Show: Next: Eq, Artifact <: AnyRef: Show: Eq](
+    stateUpdateSemaphore: Semaphore[F],
+    lastKeyAndArtifactR: Ref[F, Option[(Key, Artifact)]],
+    eventsR: MapRef[F, PeerId, Option[List[(Ordinal, Event)]]],
+    statesR: MapRef[F, Key, Option[ConsensusState[Key, Artifact]]],
+    resourcesR: MapRef[F, Key, Option[ConsensusResources[Artifact]]]
+  ): ConsensusStorage[F, Event, Key, Artifact] =
+    new ConsensusStorage[F, Event, Key, Artifact] {
+
+      private val logger = Slf4jLogger.getLogger[F]
+
+      def getState(key: Key): F[Option[ConsensusState[Key, Artifact]]] =
+        statesR(key).get
+
+      def getResources(key: Key): F[Option[ConsensusResources[Artifact]]] =
+        resourcesR(key).get
+
+      def updateState(key: Key)(stateUpdateFn: StateUpdateFn): F[Option[ConsensusState[Key, Artifact]]] =
+        stateUpdateSemaphore.permit.use { _ =>
+          for {
+            (maybeState, setter) <- statesR(key).access
+            result <- stateUpdateFn(maybeState, setter)
+            maybeNewState <- result.flatTraverse {
+              case (true, newState) =>
+                logger.trace(s"Consensus state for key ${key.show} transitioned to ${newState.show}") >>
+                  Applicative[F].pure(newState.some)
+              case (false, newState) =>
+                /** This should never occur as the access to the `ConsensusState` is protected by the semaphore */
+                logger.error(s"Consensus state for key ${key.show} failed to transition to ${newState.show}") >>
+                  Applicative[F].pure(none[ConsensusState[Key, Artifact]])
+            }
+          } yield maybeNewState
+        }
```

Comment:
I know that it may look strange, but I couldn't find any better way of dealing with the issue that I have there and the issue itself is quite complex. I will try to show the issue here, but if it will remain unclear we can have a session where it might be easier to explain it.

* Semaphore and the `access` are needed because there is at least one effectful computation inside of the update function, that cannot be moved outside of it  - pulling events from the storage in order to create the proposal.
  * If I use just the `access` without the semaphore, and there is more than one fiber that will do some effectful computations, I won't be able to revert these effects if their fiber fail to `set` the value (because only a single fiber will be able to successfully `set` the value).
  * If I use just the `semaphore` without the `access` I won't be able to perform the effects inside of the `.modify` function.
* The optional result`maybeNewState` is not actually needed but I added anyway it to lower the amount of locking with `.acquire` (in my case with `.permit`). There are 2 checks that decide if a state update is going to happen. The first check is happening before any call to `.permit`, and it's the call to `isStateUpdateRequired(state, resources)` in the `ConsensusManager`. Only if that will pass we then acquire the lock with `.permit` and then do the 2nd final check (we check again because before the lock the state may have been already updated by some other fiber). The 2nd check are the calls to: `state.facilitators.traverse(resources.peerDeclarations.get)` happening for each `case` in the big pattern-mach in `internalTryAdvanceConsensus`. This 2nd check returns `None` if the condition of having all necessary `peerDeclarations` isn't met, effectively skipping doing any effecful computations and also any state updates.

Reply:
Can you simplify it and wrap a "thunk" function into a state update transformation? In a similar way as NodeState is being changed. It will both simplify complexity and readability, and you'll get rid of semaphore locking.


```scala
stateTransition(from: ProposalMade, through: SelectingMajority, to: MajoritySelected) {
  ...
}
```

If there will be a second thread trying to make the same state shift, it will fail because of not-matching initial state:
```
threadA: from: ProposalMade -> SelectingMajority
threadB: from: ProposalMade, actual: SelectingMajority -> ...
```

Reply:
I will follow it up in the next PR. We agreed that we can merge this PR with the `Semaphore` in place as it is and the non-blocking solution will come up soon.